File: Maximum Number of coins[14-Mar-2023]/c++/memoization.cpp
class Solution{
    vector<vector<int>> dp;

    public:
        int solve(int i, vector<int>& arr, int j)
        {
            if(i > j) return 0;
            
            if(dp[i][j] != -1) return dp[i][j];
            
            int max_cost = INT_MIN;
            
            for(int b = i; b <= j; b++)
            {
                int cost = arr[b] * arr[i-1] * arr[j+1] + solve(b+1, arr, j) + solve(i, arr, b-1);
                max_cost = max(max_cost, cost);
            }
            
            return dp[i][j] = max_cost;
        }

        int maxCoins(int N, vector <int> &arr)
        {
              dp.resize(N+2, vector<int>(N+2, -1));
                arr.insert(arr.begin(), 1);
                arr.push_back(1);
                
                return solve(1, arr, N);
        }
};

File: Maximum Number of coins[14-Mar-2023]/java/memoization.java


//User function Template for Java

class Solution{
    Integer[][] dp;
    int solve(int i,ArrayList<Integer> arr,int j)
    {
        if(i>j) return 0;
        
        if(dp[i][j]!=null) return dp[i][j];
        
        int max=Integer.MIN_VALUE;
        
        for(int b=i;b<=j;b++)
        {
            int cost=arr.get(b)*arr.get(i-1)*arr.get(j+1)+solve(b+1,arr,j)+solve(i,arr,b-1);
            max=Math.max(max,cost);
        }
        
            return dp[i][j]=max;
    }
    int maxCoins(int N, ArrayList<Integer> arr) {
		
		dp=new Integer[N+2][N+2];
		
		arr.add(0,1); arr.add(1);
		
		    return solve(1,arr,N);
		
	}
}